client/ClientDecisionService/InMemoryLogger.cs

```diff
@@ -0,0 +1,237 @@
+            public Interaction InteractData { get; set; }
+            public float Reward { get; set; }
+            // TODO: This can be used to support custom reward functions
+            public ConcurrentBag<object> Outcomes = new ConcurrentBag<object>();
+            // Used to control expiration for fixed-duration events
+            public DateTime ExpiresAt = DateTime.MaxValue;
+        }
+
+        private bool disposed = false;
+
+        // The experimental unit duration, or how long to wait for reward information before 
+        // completing an event (TimeSpan.MaxValue means wait forever)
+        private TimeSpan experimentalUnit;
+        // Stores pending (incomplete) events, either for a fixed experimental unit duration or for 
+        // manual completion
+        private ConcurrentDictionary<string, DataPoint> pendingData;
+        // A queue and timer to expire events based on the experimental unit duration
+        private ConcurrentQueue<DataPoint> completionQueue;
+        private Timer completionTimer;
+        // Stores expired or manually-completed events 
+        private ConcurrentDictionary<string, DataPoint> completeData = new ConcurrentDictionary<string, DataPoint>();
+        // If no reward information is received, this value will be used
+        private float defaultReward;
+
+        /// <summary>
+        /// Creates a new in-memory logger for exploration data
+        /// </summary>
+        /// <param name="expUnit">The experimental unit duration, or how long to wait for reward 
+        /// information. Set this to TimeSpan.MaxValue for infinite duration (events never expire
+        /// and must be completed manually.</param>
+        /// <param name="defaultReward">Reward value to use when no reward signal is received</param>
+        public InMemoryLogger(TimeSpan expUnit, float defaultReward = (float)0.0)
+        {
+            this.experimentalUnit = expUnit;
+            this.defaultReward = defaultReward;
+            pendingData = new ConcurrentDictionary<string,DataPoint>();
+            // We only need the completion queue/timer if events are being completed automatically
+            // (by experimental duration)
+            if (experimentalUnit != TimeSpan.MaxValue)
+            {
+                completionQueue = new ConcurrentQueue<DataPoint>();
+                completionTimer = new Timer(experimentalUnit.TotalMilliseconds);
+                completionTimer.Elapsed += completeExpiredEvents;
+                completionTimer.AutoReset = false;
+                completionTimer.Start();
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        public void Dispose(bool disposing)
+        {
+            if (disposed) { return; }
+
+            if (disposing)
+            {
+                if (completionTimer != null)
+                {
+                    completionTimer.Dispose();

```

Comment:
Sorry, fixed. Too many chopped up pieces and fixes over the course of this PR -- next one will be cleaner and hopefully one-shot.